### DISCUSSION TITLE
Why does proxy() internally create a copy of the object?

### DISCUSSION THREAD

**Discussion by pastelmind** [pmndrs/valtio#756] [Q&A] [answered]
I'm referring to this:

https://github.com/pmndrs/valtio/blob/a828b04e05d4d7f9e2bcc00ec593142ca85fe411/src/vanilla.ts#L244-L246

Why does Valtio create a copy of the original object passed to `proxy()`? Not trying to argue, just curious about the design decision.

cf. [Vue 3's reactivity system simply wraps the Proxy over the original object](https://github.com/vuejs/core/blob/3decc57d0c0b31276ac09b3f12c1e2fc192520c1/packages/reactivity/src/reactive.ts#L279-L282).

**Comment by pastelmind** [comment]
Is it because Valtio wants to handle non-configurable, non-writable properties?

FWIW, [Vue 3 can't handle them](https://github.com/vuejs/core/issues/3024) because they wrap the target object directly.

**Comment by dai-shi** [comment, marked as answer]
It's the design choice.
We assume the initial object as immutable objects, and do not modify it. So that `proxy()` is being a pure function.
We've discussed this somewhere else, and because this behavior isn't configurable, I'm planning to change the design in the next major version: #703

**Comment by orange4glace** [comment]
I'm not fully understand about valtio but I thought that valtio is superset of `Proxy` of ES6 when I first saw.
The `Proxy` uses the original object and it is meant-to-be used to mutable object. For example,
```js
const origin = {};
const proxy = new Proxy(origin, {});
proxy.x = 10;
origin.x === 10 // true
```
From that point of view, I think using the initial object makes more sense, IMHO.
